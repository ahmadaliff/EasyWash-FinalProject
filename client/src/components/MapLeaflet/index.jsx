import { PropTypes } from 'prop-types';
import { Circle, MapContainer, TileLayer, Tooltip } from 'react-leaflet';
import { useCallback, useEffect, useState } from 'react';

import MarkerLeaflet from '@components/MarkerLeaflet';

import classes from '@components/MapLeaflet/style.module.scss';
import 'leaflet/dist/leaflet.css';
import { IconButton } from '@mui/material';
import { MyLocation } from '@mui/icons-material';
import { FormattedMessage } from 'react-intl';

const MapLeaflet = ({ handleLocation, islocated }) => {
  const [markerloc, setMarkerloc] = useState(null);
  const [myLocation, setMyLocation] = useState(null);
  const [map, setMap] = useState(null);

  useEffect(() => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition((position) => {
        const { latitude, longitude } = position.coords;
        setMyLocation({ lat: latitude, lng: longitude });
        setMarkerloc({ lat: latitude, lng: longitude });
        if (islocated) {
          setMarkerloc(islocated);
        }
        handleLocation({ lat: latitude, lng: longitude });
      });
    }
  }, []);

  const handleToMyLocation = useCallback(() => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition((position) => {
        const { latitude, longitude } = position.coords;
        setMyLocation({ lat: latitude, lng: longitude });
        map.setView({ lat: latitude, lng: longitude }, 17);
      });
    }
  }, [map]);

  return (
    <div className={classes.mapWrap}>
      {markerloc != null && (
        <MapContainer center={markerloc} zoom={17} className={classes.mapContainer} ref={setMap}>
          <TileLayer url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" />
          <Circle center={myLocation} pathOptions={{ fillColor: 'blue' }} radius={5}>
            <Tooltip sticky>
              <FormattedMessage id="app_my_location" />
            </Tooltip>
          </Circle>
          <MarkerLeaflet positionProps={markerloc} handleLocation={handleLocation} />
        </MapContainer>
      )}
      <IconButton onClick={handleToMyLocation} size="small" className={classes.myLocationButton}>
        <MyLocation />
      </IconButton>
    </div>
  );
};

MapLeaflet.propTypes = {
  handleLocation: PropTypes.func,
  islocated: PropTypes.object,
};

export default MapLeaflet;
