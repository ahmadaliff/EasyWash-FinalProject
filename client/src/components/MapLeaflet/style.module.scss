@import '@styles/base';
.mapWrap {
  position: relative;
  height: auto;
  width: 100%;
  .myLocationButton {
    position: absolute;
    bottom: 1.25rem;
    right: 0.25rem;
    z-index: 5;
    background-color: var(--color-bg);
    &:hover {
      background-color: rgba($color: gray, $alpha: 0.5);
    }
    svg {
      width: 1rem;
      height: 1rem;
    }
  }
  .mapContainer {
    position: relative;
    z-index: 4;
    height: 10rem;
    width: 100%;
  }
  .mapContainer.merchants {
    height: var(--min-height);
    @include apply-at-sm-size {
      height: calc(100vh - 6rem);
    }
  }
  .avatarMarker {
    width: 1.5rem;
    height: 1.5rem;
    box-shadow: rgba(99, 99, 99, 0.2) 0px 1px 10px 0px;
    margin-right: 0.25rem;
    background-color: var(--color-primary);
    overflow: hidden;
    img {
      width: 1.5rem;
      height: 1.5rem;
      object-fit: cover;
    }
    p {
      text-transform: uppercase;
      font-size: small;
      font-weight: 700;
      color: white;
    }
  }
}
