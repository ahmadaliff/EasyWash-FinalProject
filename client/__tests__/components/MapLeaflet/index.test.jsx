import MapLeaflet from '@components/MapLeaflet';
import { fireEvent, render } from '@utils/testHelper';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useRef: () => jest.fn(() => ({ current: {} })),
}));

jest.mock('react-leaflet', () => ({
  Circle: jest.fn().mockImplementation(({ children }) => <div>{children}</div>),
  Marker: jest.fn().mockImplementation(({ children }) => <div>{children}</div>),
  Tooltip: jest.fn().mockImplementation(({ children }) => <div>{children}</div>),
  TileLayer: jest.fn().mockImplementation(({ children }) => <div>{children}</div>),
  MapContainer: jest.fn().mockImplementation(({ children }) => <div>{children}</div>),
  useMapEvents: () => jest.fn(),
}));

jest.mock('@react-leaflet/core', () => ({ createControlComponent: () => jest.fn() }));

jest.mock('react-dom/server', () => ({
  renderToStaticMarkup: jest.fn(),
}));

jest.mock('react-leaflet-cluster', () => jest.fn().mockImplementation(({ children }) => <div>{children}</div>));

jest.mock('@components/MarkerLeaflet', () => () => <div>MockMarker</div>);

const mockGeolocation = {
  getCurrentPosition: jest.fn().mockImplementationOnce((success) =>
    Promise.resolve(
      success({
        coords: {
          latitude: -6.224934545453234,
          longitude: 106.86266439802276,
        },
      })
    )
  ),
};
global.navigator.geolocation = mockGeolocation;

let wrapper;

const mockProps = {
  handleLocation: jest.fn(),
  isLocated: { lat: -6.224934545453234, lng: 106.86266439802276 },
  login: true,
  user: {
    fullName: 'ahmad alif sofian',
    role: 'user',
    imagePath: null,
  },
};

describe('Map Leaflet Component', () => {
  describe('map leaflet wihout merchants props', () => {
    beforeEach(() => {
      jest.spyOn(console, 'error').mockImplementation(() => {});
      wrapper = render(<MapLeaflet {...mockProps} />);
      // eslint-disable-next-line no-console
      console.error.mockRestore();
    });
    test('Correct render', () => {
      const { getByTestId } = wrapper;
      expect(getByTestId('map-wrap')).toBeInTheDocument();
    });
    test('Correct render button to get current position', () => {
      const { getByTestId } = wrapper;
      expect(getByTestId('button-position')).toBeInTheDocument();
      const button = getByTestId('button-position');
      fireEvent.click(button);
    });
    test('Should match with snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
