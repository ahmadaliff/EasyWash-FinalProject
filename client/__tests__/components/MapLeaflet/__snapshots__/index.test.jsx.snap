// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Leaflet Component map leaflet wihout merchants props Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mapWrap"
        data-testid="map-wrap"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall myLocationButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
          data-testid="button-position"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MyLocationIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mapWrap"
      data-testid="map-wrap"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall myLocationButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        data-testid="button-position"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MyLocationIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
